internal/pkg/deploy/cloudformation/stack/app.go

```diff
@@ -1,157 +1,243 @@
+	appOutputS3Bucket             = "PipelineBucket"
+	appOutputECRRepoPrefix        = "ECRRepo"
+	appDNSDelegatedAccountsKey    = "AppDNSDelegatedAccounts"
+	appDomainNameKey              = "AppDomainName"
+	appNameKey                    = "AppName"
+	appDNSDelegationRoleName      = "DNSDelegationRole"
 )
 
-// RuntimeConfig represents configuration that's defined outside of the manifest file
-// that is needed to create a CloudFormation stack.
-type RuntimeConfig struct {
-	ImageRepoURL      string            // ImageRepoURL is the ECR repository URL the container image should be pushed to.
-	ImageTag          string            // ImageTag is the container image's unique tag.
-	AddonsTemplateURL string            // Optional. S3 object URL for the addons template.
-	AdditionalTags    map[string]string // AdditionalTags are labels applied to resources in the application stack.
+// AppConfigFrom takes a template file and extracts the metadata block,
+// and parses it into an AppStackConfig
+func AppConfigFrom(template *string) (*AppResourcesConfig, error) {
+	resourceConfig := DeployedAppMetadata{}
+	err := yaml.Unmarshal([]byte(*template), &resourceConfig)
+	return &resourceConfig.Metadata, err
 }
 
-type templater interface {
-	Template() (string, error)
+// NewAppStackConfig sets up a struct which can provide values to CloudFormation for
+// spinning up an environment.
+func NewAppStackConfig(in *deploy.CreateProjectInput) *AppStackConfig {
+	return &AppStackConfig{
+		CreateProjectInput: in,
+		parser:             template.New(),
+	}
 }
 
-type app struct {
-	name    string
-	env     string
-	project string
-	tc      manifest.TaskConfig
-	rc      RuntimeConfig
-
-	parser template.Parser
-	addons templater
+// Template returns the environment CloudFormation template.
+func (c *AppStackConfig) Template() (string, error) {
+	content, err := c.parser.Read(appTemplatePath)
+	if err != nil {
+		return "", err
+	}
+	return content.String(), nil
 }
 
-// StackName returns the name of the stack.
-func (a *app) StackName() string {
-	return NameForApp(a.project, a.env, a.name)
+// ResourceTemplate generates a StackSet template with all the Application-wide resources (ECR Repos, KMS keys, S3 buckets)
+func (c *AppStackConfig) ResourceTemplate(config *AppResourcesConfig) (string, error) {
+	// Sort the account IDs and Services so that the template we generate is deterministic
+	sort.Strings(config.Accounts)
+	sort.Strings(config.Services)
+
+	content, err := c.parser.Parse(appResourcesTemplatePath, struct {
+		*AppResourcesConfig
+		ServiceTagKey string
+	}{
+		config,
+		ServiceTagKey,
+	}, template.WithFuncs(templateFunctions))
+	if err != nil {
+		return "", err
+	}
+	return content.String(), err
 }
 
-// Parameters returns the list of CloudFormation parameters used by the template.
-func (a *app) Parameters() []*cloudformation.Parameter {
+// Parameters returns the parameters to be passed into a environment CloudFormation template.
```

Comment:
How about something like: 
```suggestion
// Parameters returns a list of parameters which accompany the app CloudFormation template.
```